rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,218 @@
+  bytes  metadata        = 5;  // Arbitrary application metadata
+  uint64 backfill_height = 6;  // Height to backfill blocks from before starting application (inclusive)
+}
+```
+
+If all chunks are accepted and state sync is successful then the node will
+retrieve and verify blocks up to (and including) the specified backfill height
+**before** participating in consensus.
+
+### ReverseSync on application request
+
+The application has control of block retention via `retain_height`, called here:
+
+```proto
+message ResponseCommit {
+  // reserve 1
+  bytes  data          = 2;
+  uint64 retain_height = 3;
+}
+```
+
+Beforehand, a retain height that was less than the nodes current base was
+ignored. With the same backfill mechanism, the application should now be able to
+lower the retention height below the current base height (whether to accommodate
+changing consensus parameters, app specific needs, or a local config).
+This accommodates the use case of a truncated node that wishes to become an
+archive node.
+
+#### Choosing between Strong or Weak Gaurantees about Data
+
+It can be said previously, that block pruning, with respect to the application, 
+offered strong gaurantees about data i.e. if the application pruned to height 100, 
+the operation was synchronous and thus the application knew it would prune to 100 
+before the next height. This meant that the application always knew the height and 
+base of the blockchain at every step. The opposite, weak gaurantees, implies that 
+the application can't gaurantee that what it asked from Tendermint, with respect 
+to data, has happened.
+
+The downside of strong gaurantees is that it is blocking. If a validator decided 
+to prune a million blocks at once, the validator could potentially fall behind. 
+There's an argument then that pruning (and for that matter reverse sync) shouldn't 
+lie on the consensus critical path but rather be an asynchronous background process 
+of the node.
+
+Further to this argument, nodes will not be able to prune past the evidence age 
+regardless of the retain height that the application requests. Finally, applications 
+may ask to fetch blocks back to a height that none of the nodes peers have. 
+The mechanism must ensure termination even if it doesn't reach the height requested 
+by the app.
+
+Taking all these points into consideration, ReverseSync and pruning should be
+operations that run asynchronously in the background. Application can be informed
+of the current base via `RequestEndBlock`:
+
+```proto 
+message RequestEndBlock {
+  Height uint64
+  Base uint64
+}
+```
+
+### Verification
+
+Nodes will need to verify these prior blocks. This can be achieved by first
+retrieving the header at the base height from the block store. The node then
+checks that the `LastBlockID` corresponds with the hash calculated from the
+header in the new block directly below:
+
+```go
+header[height].LastBlockID == hash(header[height-1])
```

Comment:
Yeah we can. It's basically the same process of checking the LastBlockID and then once we trust the Header, verifying the rest of the data through the hashes

Reply:
There is a difference in that the light client's `VerifyBackwards` should allow for far more parallelism in verifying / checking old blocks, whereas this method enforces a sequential check before you can ensure you've been given a valid header. This should matter in terms of bounding how much garbage data an adversarial peer can give you right?

Reply:
Although it's possible to do it in a skipping manner (i.e `VerifySkipping` but backwards), `VerifyBackwards` currently only verifies sequentially which means that there is no parallelism and that it must step through every header. Thus, these algorithms are basically the same.

In case your wondering, we never got around to implementing the verify backwards skipping algorithm because we figured that most light clients would start on a trusted light block and naturally keep verifying forwards. There's probably an open issue for this but perhaps `VerifyBackwardsSkipping` is something that we can add to a future light client

Reply:
That makes sense, didn't realize `VerifyBackwardsSkipping` is a TODO atm.

Reply:
How you can implement VerifyBackwardsSkipping? My understanding was that backward verification is strictly sequential as we need to follow the hash links.

Reply:
My thinking is that it should be possible but I could be wrong.

Say you wanted to verify backwards from height 100 to height 10 (Both of them are within the trusting period). If 1/3+ of the validators in height 100 were also present in the validator set at height 10 and signed the header at that height we should be able to trust the header in the same way that we can trust via skipping verification. We would then still run header comparison with witnesses. If 1/3+ had performed a lunatic attack and a witness was able to produce a conflicting header then we would create evidence and slash the common validators that were in height 100 and signed in height 10